src/python/pants/backend/python/subsystems/setup.py

```diff
@@ -111,6 +125,20 @@ def interpreter_constraints(self) -> tuple[str, ...]:
                     """
                 ),
             )
+
+        # Warn if Python 2.x is still in use. This warning should only be displayed once since this
+        # function is memoized.
+        if self.warn_on_python2_usage:
```

Comment:
It looks like this is only detecting ICs that are configured in `[python].interpreter_constraints = ...` but I think there's other routes that ICs can be set, e.g. the `[python].resolves_to_interpreter_constraints` option in this file, or [`interpreter_constraints` fields](https://www.pantsbuild.org/stable/reference/targets/pex_binary#interpreter_constraints) on individual targets.

It'd be good to detect those too, which might mean pushing the warning down lower?

(I think `pants.base.deprecated.warn_or_error` helps with the memorisation too.)

Reply:
I had looked at `pants.base.deprecated.warn_or_error` but it can raise an exception so seems inapplicable to this soft-deprecation.

Reply:
Ah right, we're not removing any active support code once this deprecation "expires", just stopping the testing. So yeah, ignore the `warn_or_error` suggestion 😄

Reply:
And the memoization comes from `warn_or_error` being marked with @memoized. Could do something similar with `warn_on_python2_usage_in_interpreter_constraints` since it includes `description_of_origin` which can differentiate call sites.

Reply:
> but I think there's other routes that ICs can be set, e.g. the [python].resolves_to_interpreter_constraints option in this file, or [interpreter_constraints fields](https://www.pantsbuild.org/stable/reference/targets/pex_binary#interpreter_constraints) on individual targets

There are many calls to the `InterpreterConstraints` constructor as well.

Maybe just checks on the field and the `--python-resolves-to-interpreter-constraints` option? Latest commit adds that.